File: weird_captcha_gym/shared_runtime/app/mechanics/velvet_valet.js
(() => {
  "use strict";
  let model = null;

  function corners(x, y, length, width, angle) {
    const ca = Math.cos(angle), sa = Math.sin(angle);
    return [[-length / 2, -width / 2], [length / 2, -width / 2], [length / 2, width / 2], [-length / 2, width / 2]]
      .map(([dx, dy]) => [x + ca * dx - sa * dy, y + sa * dx + ca * dy]);
  }
  function axes(points) {
    const result = [];
    for (let i = 0; i < points.length; i += 1) {
      const a = points[i], b = points[(i + 1) % points.length];
      const dx = b[0] - a[0], dy = b[1] - a[1], length = Math.hypot(dx, dy);
      if (length > 1e-9) result.push([-dy / length, dx / length]);
    }
    return result;
  }
  function overlaps(first, second) {
    for (const [ax, ay] of axes(first).concat(axes(second))) {
      const a = first.map(point => point[0] * ax + point[1] * ay);
      const b = second.map(point => point[0] * ax + point[1] * ay);
      if (Math.max(...a) <= Math.min(...b) + 1e-7 || Math.max(...b) <= Math.min(...a) + 1e-7) return false;
    }
    return true;
  }
  function wrapAngle(value) { return ((value + Math.PI) % (2 * Math.PI) + 2 * Math.PI) % (2 * Math.PI) - Math.PI; }
  function angleError(a, b) { return Math.abs(wrapAngle(a - b)); }
  function setControl(command, inputSource) {
    if (!model || model.done) return;
    applyCommand(command);
    model.actions.push({sequence: model.actions.length + 1, tick: model.tick, type: "control", command, input_source: inputSource});
    model.helpers.setReadout("", "idle");
  }
  function applyCommand(command) {
    if (command === "steer_left") model.controls.steer = -1;
    else if (command === "steer_right") model.controls.steer = 1;
    else if (command === "steer_center") model.controls.steer = 0;
    else if (command === "forward") { model.controls.throttle = 1; model.controls.brake = false; }
    else if (command === "reverse") { model.controls.throttle = -1; model.controls.brake = false; }
    else if (command === "coast") model.controls.throttle = 0;
    else if (command === "brake_on") { model.controls.brake = true; model.controls.throttle = 0; }
    else if (command === "brake_off") model.controls.brake = false;
  }
  function carCorners() {
    const physics = model.state.world.physics;
    return corners(model.car.x, model.car.y, physics.car_length, physics.car_width, model.car.heading);
  }
  function collisionReason() {
    const world = model.state.world, physics = world.physics, car = carCorners();
    if (car.some(([x, y]) => x < 0 || x > Number(world.width) || y < 0 || y > Number(world.height))) return "KERB COLLISION";
    for (const obstacle of world.obstacles) {
      const other = corners(Number(obstacle.x), Number(obstacle.y), Number(obstacle.width), Number(obstacle.height), Number(obstacle.angle || 0));
      if (overlaps(car, other)) return `COLLISION — ${obstacle.id.toUpperCase()}`;
    }
    return null;
  }
  function advance() {
    const physics = model.state.world.physics;
    let speed = model.car.speed;
    if (model.controls.brake) speed *= Number(physics.brake_factor);
    else if (model.controls.throttle > 0) speed = Math.min(Number(physics.max_forward_speed), speed + Number(physics.acceleration));
    else if (model.controls.throttle < 0) speed = Math.max(-Number(physics.max_reverse_speed), speed - Number(physics.reverse_acceleration));
    else speed *= Number(physics.coast_factor);
    const fraction = Math.min(1, Math.abs(speed) / Math.max(Number(physics.max_forward_speed), 0.001));
    const direction = speed < 0 ? -1 : 1;
    model.car.heading = wrapAngle(model.car.heading + Number(model.controls.steer) * Number(physics.turn_rate) * (0.25 + 0.75 * fraction) * direction);
    model.car.x += Math.cos(model.car.heading) * speed;
    model.car.y += Math.sin(model.car.heading) * speed;
    model.car.speed = speed;
    model.tick += 1;
    return collisionReason();
  }
  function draw() {
    if (!model) return;
    const world = model.state.world, car = document.querySelector(".vv-car");
    if (car) {
      car.style.left = `${model.car.x / Number(world.width) * 100}%`;
      car.style.top = `${model.car.y / Number(world.height) * 100}%`;
      car.style.transform = `translate(-50%, -50%) rotate(${model.car.heading * 180 / Math.PI}deg)`;
      car.dataset.x = model.car.x.toFixed(3); car.dataset.y = model.car.y.toFixed(3);
      car.dataset.heading = model.car.heading.toFixed(5); car.dataset.speed = model.car.speed.toFixed(4);
    }
    const tick = document.querySelector(".vv-tick b"), speed = document.querySelector(".vv-speed b"), pose = document.querySelector(".vv-pose b");
    if (tick) tick.textContent = String(model.tick).padStart(4, "0");
    if (speed) speed.textContent = `${model.car.speed.toFixed(2)} px/tick`;
    if (pose) pose.textContent = `${Math.round(model.car.x)}, ${Math.round(model.car.y)} · ${Math.round(model.car.heading * 180 / Math.PI)}°`;
  }
  function parked() {
    const world = model.state.world, physics = world.physics, target = world.target;
    if (Math.hypot(model.car.x - Number(target.x), model.car.y - Number(target.y)) > Number(physics.position_tolerance)) return false;
    if (angleError(model.car.heading, Number(target.heading)) > Number(physics.heading_tolerance_deg) * Math.PI / 180) return false;
    if (Math.abs(model.car.speed) > Number(physics.speed_tolerance)) return false;
    const ca = Math.cos(Number(target.heading)), sa = Math.sin(Number(target.heading));
    return carCorners().every(([x, y]) => {
      const localX = (x - Number(target.x)) * ca + (y - Number(target.y)) * sa;
      const localY = -(x - Number(target.x)) * sa + (y - Number(target.y)) * ca;
      return Math.abs(localX) <= Number(target.width) / 2 && Math.abs(localY) <= Number(target.length) / 2;
    });
  }
  async function finish(completed, reason) {
    if (!model || model.done) return;
    model.done = true; clearInterval(model.timer);
    model.helpers.setReadout(reason === "park" ? "CHECKING PARKED POSE…" : reason, reason === "park" ? "busy" : "error");
    try {
      const response = await fetch("/result", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify({
        mechanic_id: model.state.mechanic_id, task_id: model.state.task_id, challenge_id: model.state.challenge_id,
        events: model.actions, final_tick: model.tick, completed, reason,
      })});
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.helpers.setReadout("PARKED · PASS", "passed");
        document.querySelector(".vv")?.classList.add("is-passed");
      } else {
        model.helpers.setReadout(`FAIL · ${outcome.feedback || reason}`, "error");
        document.querySelector(".vv")?.classList.add("is-failed");
        if (outcome.state) setTimeout(() => model.helpers.render(outcome.state), 850);
      }
    } catch (_error) {
      model.done = false; model.helpers.setReadout("RESULT CONNECTION FAILED", "error");
    }
  }
  function controlMarkup(interaction) {
    if (interaction === "simplified") return `<div class="vv-controls vv-buttons" aria-label="valet proxy controls">
      <button data-vv-command="steer_left">TURN LEFT</button><button data-vv-command="steer_center">WHEELS STRAIGHT</button><button data-vv-command="steer_right">TURN RIGHT</button>
      <button data-vv-command="forward">FORWARD</button><button data-vv-command="coast">COAST</button><button data-vv-command="reverse">REVERSE</button><button data-vv-command="brake_on">BRAKE</button>
    </div>`;
    return `<div class="vv-controls vv-key-guide"><span>FULL INPUT</span><b>↑ DRIVE</b><b>↓ REVERSE</b><b>← → STEER</b><b>SPACE BRAKE</b></div>`;
  }
  async function render(state, helpers) {
    document.body.dataset.mechanic = "velvet-valet";
    if (model?.timer) clearInterval(model.timer);
    if (model?.keyHandler) window.removeEventListener("keydown", model.keyHandler);
    if (model?.keyUpHandler) window.removeEventListener("keyup", model.keyUpHandler);
    const interaction = state.control_condition?.interaction || "full", world = state.world, physics = world.physics;
    model = {state, helpers, interaction, tick: 0, actions: [], done: false, started: false, timer: null,
      controls: {steer: 0, throttle: 0, brake: false}, car: {...world.start, speed: 0}, keyHandler: null, keyUpHandler: null};
    const obstacles = world.obstacles.map(item => `<div class="vv-obstacle shade-${item.shade}" style="left:${Number(item.x) / Number(world.width) * 100}%;top:${Number(item.y) / Number(world.height) * 100}%;width:${Number(item.width) / Number(world.width) * 100}%;height:${Number(item.height) / Number(world.height) * 100}%" data-obstacle="${item.id}"><i></i><b></b></div>`).join("");
    const bay = `<div class="vv-bay" style="left:${Number(world.target.x) / Number(world.width) * 100}%;top:${Number(world.target.y) / Number(world.height) * 100}%;width:${Number(world.target.width) / Number(world.width) * 100}%;height:${Number(world.target.length) / Number(world.height) * 100}%"><span>REVERSE IN</span><b>▼</b></div>`;
    helpers.app.innerHTML = `<section class="vv" tabindex="0"><header><div><small>THEATRE COURTYARD · NONHOLONOMIC PARKING</small><h1>Velvet Valet</h1><p>${state.prompt}</p></div><div class="vv-readouts"><span class="vv-tick">TICK <b>0000</b></span><span class="vv-speed">SPEED <b>0.00 px/tick</b></span><span class="vv-pose">POSE <b>—</b></span></div></header>
      <main class="vv-courtyard" style="--world-w:${world.width};--world-h:${world.height}"><div class="vv-kerb vv-kerb-top"></div><div class="vv-kerb vv-kerb-bottom"></div><div class="vv-kerb vv-kerb-left"></div><div class="vv-kerb vv-kerb-right"></div>${bay}${obstacles}<div class="vv-car" data-x="${world.start.x}" data-y="${world.start.y}" data-heading="${world.start.heading}" data-speed="0"><i></i><b></b></div><button class="vv-start" type="button">START ENGINE</button></main>
      <footer><div class="readout vv-status" data-status="idle"></div><button class="vv-submit" type="button" disabled>CHECK PARKING</button>${controlMarkup(interaction)}</footer></section>`;
    const shell = document.querySelector(".vv"), startButton = document.querySelector(".vv-start"), submit = document.querySelector(".vv-submit");
    function start() { if (model.started || model.done) return; model.started = true; startButton.remove(); submit.disabled = false; shell.focus(); model.timer = setInterval(step, Number(physics.tick_ms)); model.helpers.setReadout("ENGINE RUNNING", "idle"); }
    function step() { if (!model || model.done || !model.started) return; const collision = advance(); draw(); if (collision) finish(false, collision); else if (model.tick >= Number(physics.max_ticks)) finish(false, "TIME LIMIT"); }
    model.keyHandler = event => {
      if (interaction !== "full" || event.repeat) return;
      const map = {ArrowLeft: "steer_left", ArrowRight: "steer_right", ArrowUp: "forward", ArrowDown: "reverse", " ": "brake_on"};
      if (!map[event.key]) return; event.preventDefault(); if (!model.started) start(); setControl(map[event.key], "keyboard");
    };
    model.keyUpHandler = event => {
      if (interaction !== "full") return;
      const map = {ArrowLeft: "steer_center", ArrowRight: "steer_center", ArrowUp: "coast", ArrowDown: "coast", " ": "brake_off"};
      if (!map[event.key]) return; event.preventDefault(); setControl(map[event.key], "keyboard");
    };
    window.addEventListener("keydown", model.keyHandler); window.addEventListener("keyup", model.keyUpHandler);
    startButton.addEventListener("click", start);
    document.querySelectorAll("[data-vv-command]").forEach(button => button.addEventListener("click", () => { if (!model.started) start(); setControl(button.dataset.vvCommand, "control_button"); }));
    submit.addEventListener("click", () => { if (model.started && !model.done) finish(parked(), "park"); });
    shell.focus(); draw();
  }
  function step() {}
  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.velvet_valet = {render, rootSelector: ".vv"};
})();

File: weird_captcha_gym/shared_runtime/server/incubator_graders/velvet_valet.py
from __future__ import annotations

import math
from typing import Any


MECHANIC_ID = "velvet_valet"


def _fail(message: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "feedback": message}


def _angle_error(first: float, second: float) -> float:
    return abs((first - second + math.pi) % (2 * math.pi) - math.pi)


def _corners(x: float, y: float, length: float, width: float, angle: float) -> list[tuple[float, float]]:
    ca, sa = math.cos(angle), math.sin(angle)
    return [
        (x + ca * dx - sa * dy, y + sa * dx + ca * dy)
        for dx, dy in ((-length / 2, -width / 2), (length / 2, -width / 2),
                       (length / 2, width / 2), (-length / 2, width / 2))
    ]


def _axes(points: list[tuple[float, float]]) -> list[tuple[float, float]]:
    axes = []
    for index, first in enumerate(points):
        second = points[(index + 1) % len(points)]
        dx, dy = second[0] - first[0], second[1] - first[1]
        length = math.hypot(dx, dy)
        if length > 1e-9:
            axes.append((-dy / length, dx / length))
    return axes


def _overlap(first: list[tuple[float, float]], second: list[tuple[float, float]]) -> bool:
    for axis in _axes(first) + _axes(second):
        a = [point[0] * axis[0] + point[1] * axis[1] for point in first]
        b = [point[0] * axis[0] + point[1] * axis[1] for point in second]
        if max(a) <= min(b) + 1e-7 or max(b) <= min(a) + 1e-7:
            return False
    return True


def _state_valid(state: dict[str, float], world: dict[str, Any]) -> str | None:
    physics = world["physics"]
    car = _corners(
        float(state["x"]), float(state["y"]), float(physics["car_length"]),
        float(physics["car_width"]), float(state["heading"]),
    )
    if any(point[0] < 0 or point[0] > float(world["width"]) or point[1] < 0 or point[1] > float(world["height"]) for point in car):
        return "kerb collision"
    for obstacle in world["obstacles"]:
        other = _corners(
            float(obstacle["x"]), float(obstacle["y"]), float(obstacle["width"]),
            float(obstacle["height"]), float(obstacle.get("angle", 0.0)),
        )
        if _overlap(car, other):
            return f"collision with {obstacle['id']}"
    return None


def _advance(state: dict[str, float], controls: dict[str, Any], world: dict[str, Any]) -> str | None:
    physics = world["physics"]
    throttle = int(controls.get("throttle", 0))
    speed = float(state["speed"])
    if bool(controls.get("brake")):
        speed *= float(physics["brake_factor"])
    elif throttle > 0:
        speed = min(float(physics["max_forward_speed"]), speed + float(physics["acceleration"]))
    elif throttle < 0:
        speed = max(-float(physics["max_reverse_speed"]), speed - float(physics["reverse_acceleration"]))
    else:
        speed *= float(physics["coast_factor"])
    steer = max(-1, min(1, int(controls.get("steer", 0))))
    speed_fraction = min(1.0, abs(speed) / max(float(physics["max_forward_speed"]), 0.001))
    direction = -1.0 if speed < 0 else 1.0
    state["heading"] = (float(state["heading"]) + steer * float(physics["turn_rate"]) * (0.25 + 0.75 * speed_fraction) * direction + math.pi) % (2 * math.pi) - math.pi
    state["x"] += math.cos(float(state["heading"])) * speed
    state["y"] += math.sin(float(state["heading"])) * speed
    state["speed"] = speed
    return _state_valid(state, world)


def _apply_command(controls: dict[str, Any], command: str) -> None:
    if command == "steer_left":
        controls["steer"] = -1
    elif command == "steer_right":
        controls["steer"] = 1
    elif command == "steer_center":
        controls["steer"] = 0
    elif command == "forward":
        controls["throttle"] = 1
        controls["brake"] = False
    elif command == "reverse":
        controls["throttle"] = -1
        controls["brake"] = False
    elif command == "coast":
        controls["throttle"] = 0
    elif command == "brake_on":
        controls["brake"] = True
        controls["throttle"] = 0
    elif command == "brake_off":
        controls["brake"] = False
    else:
        raise ValueError(f"unknown control command {command!r}")


def _parked(state: dict[str, float], world: dict[str, Any]) -> tuple[bool, str]:
    physics, target = world["physics"], world["target"]
    distance = math.hypot(float(state["x"]) - float(target["x"]), float(state["y"]) - float(target["y"]))
    position_limit = float(physics["position_tolerance"])
    heading_limit = math.radians(float(physics["heading_tolerance_deg"]))
    if distance > position_limit:
        return False, f"bay center error {distance:.1f}px"
    if _angle_error(float(state["heading"]), float(target["heading"])) > heading_limit:
        return False, "nose is not aligned with the bay arrow"
    if abs(float(state["speed"])) > float(physics["speed_tolerance"]):
        return False, "car is still moving"
    car = _corners(float(state["x"]), float(state["y"]), float(physics["car_length"]), float(physics["car_width"]), float(state["heading"]))
    ca, sa = math.cos(float(target["heading"])), math.sin(float(target["heading"]))
    for px, py in car:
        local_x = (px - float(target["x"])) * ca + (py - float(target["y"])) * sa
        local_y = -(px - float(target["x"])) * sa + (py - float(target["y"])) * ca
        if abs(local_x) > float(target["width"]) / 2 or abs(local_y) > float(target["length"]) / 2:
            return False, "car footprint crosses the velvet bay border"
    return True, "car is stationary inside the velvet bay"


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    if any(source.get("mechanic_id") != MECHANIC_ID for source in (payload, ground_truth, public_state)):
        return _fail("mechanic mismatch")
    challenge_id = str(ground_truth.get("challenge_id") or "")
    if not challenge_id or payload.get("challenge_id") != challenge_id or public_state.get("challenge_id") != challenge_id:
        return _fail("stale courtyard challenge")
    if payload.get("task_id") != ground_truth.get("task_id") or public_state.get("task_id") != ground_truth.get("task_id"):
        return _fail("stale task identity")
    if public_state.get("world") != ground_truth.get("world"):
        return _fail("visible courtyard differs from the replay world")
    condition = ground_truth.get("control_condition")
    if public_state.get("control_condition") != condition:
        return _fail("interaction or difficulty condition differs from the visible contract")
    interaction = str((condition or {}).get("interaction") or "")
    expected_source = {"simplified": "control_button", "full": "keyboard"}.get(interaction)
    if condition is not None and expected_source is None:
        return _fail("invalid interaction condition")
    world = ground_truth.get("world")
    if not isinstance(world, dict):
        return _fail("missing courtyard world")
    try:
        physics = world["physics"]
        max_ticks = int(physics["max_ticks"])
        events = payload.get("events")
        final_tick = int(payload.get("final_tick"))
    except (KeyError, TypeError, ValueError):
        return _fail("parking transcript is malformed")
    if not isinstance(events, list) or len(events) > 2400 or not 0 <= final_tick <= max_ticks:
        return _fail("parking transcript is outside supported limits")
    start = world["start"]
    state = {"x": float(start["x"]), "y": float(start["y"]), "heading": float(start["heading"]), "speed": 0.0}
    controls: dict[str, Any] = {"steer": 0, "throttle": 0, "brake": False}
    tick = 0
    collision: str | None = None
    for sequence, event in enumerate(events, 1):
        if not isinstance(event, dict) or event.get("sequence") != sequence:
            return _fail(f"event {sequence} sequence mismatch")
        if event.get("input_source") != expected_source:
            return _fail(f"event {sequence} uses the wrong interaction input")
        try:
            event_tick = int(event["tick"])
        except (KeyError, TypeError, ValueError):
            return _fail(f"event {sequence} has no valid tick")
        if event_tick < tick or event_tick > final_tick:
            return _fail(f"event {sequence} is out of chronological order")
        while tick < event_tick:
            collision = _advance(state, controls, world)
            tick += 1
            if collision:
                return _fail(collision)
        if event.get("type") != "control":
            return _fail(f"event {sequence} is not a control event")
        try:
            _apply_command(controls, str(event["command"]))
        except ValueError as exc:
            return _fail(str(exc))
    while tick < final_tick:
        collision = _advance(state, controls, world)
        tick += 1
        if collision:
            return _fail(collision)
    accepted, feedback = _parked(state, world)
    passed = payload.get("completed") is True and accepted and not collision
    return {
        "graded": True,
        "passed": passed,
        "feedback": feedback if passed else feedback,
        "replay": {
            "final_tick": tick,
            "x": round(state["x"], 3),
            "y": round(state["y"], 3),
            "heading": round(state["heading"], 5),
            "speed": round(state["speed"], 4),
            "accepted": accepted,
        },
    }
